cmd/gardenlet/app/gardenlet.go

```diff
@@ -0,0 +1,453 @@
+		cfg.GardenClientConnection.Kubeconfig = kubeconfig
+	}
+	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
+		cfg.SeedClientConnection.Kubeconfig = kubeconfig
+	}
+
+	// Get a kubeconfig via CSR
+	kubeconfigFromBootstrap, err := bootstrapKubeconfig(ctx, logger, cfg.GardenClientConnection, cfg.SeedClientConnection, cfg.SeedConfig)
+	if err != nil {
+		return nil, err
+	}
+	if kubeconfigFromBootstrap == nil {
+		logger.Info("No kubeconfig generated from bootstrap process found. Using kubeconfig specified in `.gardenClientConnection` in configuration")
+	}
+
+	restCfg, err := utils.RESTConfigFromClientConnectionConfiguration(cfg.GardenClientConnection.ClientConnectionConfiguration, kubeconfigFromBootstrap)
+	if err != nil {
+		return nil, err
+	}
+
+	disc, err := discoveryFromGardenletConfiguration(cfg, kubeconfigFromBootstrap)
+	if err != nil {
+		return nil, err
+	}
+
+	k8sGardenClient, err := kubernetes.NewWithConfig(
+		kubernetes.WithRESTConfig(restCfg),
+		kubernetes.WithClientOptions(client.Options{
+			Mapper: restmapper.NewDeferredDiscoveryRESTMapper(disc),
+			Scheme: kubernetes.GardenScheme,
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	seedRestCfg, err := utils.RESTConfigFromClientConnectionConfiguration(cfg.SeedClientConnection, nil)
+	if err != nil {
+		return nil, err
+	}
+	k8sGardenClientLeaderElection, err := k8s.NewForConfig(seedRestCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up leader election if enabled and prepare event recorder.
+	var (
+		leaderElectionConfig *leaderelection.LeaderElectionConfig
+		recorder             = utils.CreateRecorder(k8sGardenClient.Kubernetes(), "gardenlet")
+	)
+	if cfg.LeaderElection.LeaderElect {
+		leaderElectionConfig, err = utils.MakeLeaderElectionConfig(cfg.LeaderElection.LeaderElectionConfiguration, cfg.LeaderElection.LockObjectNamespace, cfg.LeaderElection.LockObjectName, k8sGardenClientLeaderElection, recorder)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	identity, gardenNamespace, err := determineGardenletIdentity()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Gardenlet{
+		Identity:               identity,
+		GardenNamespace:        gardenNamespace,
+		Config:                 cfg,
+		Logger:                 logger,
+		Recorder:               recorder,
+		K8sGardenClient:        k8sGardenClient,
+		K8sGardenCoreInformers: gardencoreinformers.NewSharedInformerFactory(k8sGardenClient.GardenCore(), 0),
+		KubeInformerFactory:    kubeinformers.NewSharedInformerFactory(k8sGardenClient.Kubernetes(), 0),
+		LeaderElection:         leaderElectionConfig,
+	}, nil
+}
+
+func (g *Gardenlet) cleanup() {
+	if err := os.RemoveAll(configv1alpha1.DefaultDiscoveryDir); err != nil {
+		g.Logger.Errorf("Could not cleanup base discovery cache directory: %v", err)
+	}
+}
+
+// Run runs the Garden. This should never exit.
+func (g *Gardenlet) Run(ctx context.Context, cancel context.CancelFunc) error {
```

Comment:
Why are we using this pattern? Ideally we should not pass cancel for context as separate parameter. Cancel should always be called by context creator. We should generate new wrapper context here like leaderElection context you created below. And use that cancel function there.  
```suggestion
func (g *Gardenlet) Run(ctx context.Context) error {
 controllerCtx, controllerCancel := context.WithCancel(ctx)
```

Reply:
Hm, dunno, to be honest, I simply chose the existing code from the GCM. Eventually, we want to migrate to use the controller-runtime, so, should we keep it now as is?

Reply:
Ok Np. I am just trying to have the standard sensible usage for ctx. Said that, i think migrating to controller-runtime wont affect change here.

Reply:
Will adapt